### PR TITLE
[torch_xla2] DistributedDataParallel wrapper with minGPT example

### PR DESCRIPTION
**Pull request by will-cromar**
Implement `FSDPv2`-style DDP using JAX SPMD.

Splits inputs along batch dimension (assumed to be 0) across all devices in JAX runtime, including remote devices. Each process' shard of input data is split among the addressable devices (e.g. local TPU chips) by `shard_input`.

Example usage for eager:

```python
  jax_model = torch_xla2.distributed.DistributedDataParallel(create_model())
  for data, dataloader:
    jax_data = jax_model.shard_input(data)
    jax_output = jax_model(jax_data)
```

`jax.jit` requires that any mutated tensors are inputs and outputs of the function. Wrapping your train step in `DistributedDataParallel.jit_step` takes care of that:

```python
  @jax_model.jit_step
  def step_fn(jax_data, jax_target):
    jax_optimizer.zero_grad()
    ...
    jax_loss.backward()
    jax_optimizer.step()
```

Include new minGPT example using demo sorting dataset. I'll very likely replace this with something more realistic later.

Misc fixes
- Use `convert_dtype` on `arange`, and cast inputs using plain `numpy`. Prevents creation of invalid traced arrays when in JIT (start and stop must be constant)
- Register other variants of `bernoulli`
- Register decompositions for NLL loss
- Random trivial ops that were missing (`new_empty_strided`, `_unsafe_index_put`, `conj_physical`)

### CODE REVIEW

**Review by ManfeiBai — APPROVED** [review]
Thanks, LGTM for dtype convert of arange

**Review by qihqi — APPROVED** [review]
